File: imagine/core/norm_dist.py
import numpy as np


class NormDist:
    def __init__(self, mean=0, sd=0):
        if isinstance(mean, (int, float)) and isinstance(sd, (int, float)):
            self.mean = mean
            self.sd = sd
        else:
            raise ValueError('Can only create NormDist with numeric input.')

    @staticmethod
    def init(means, sds):
        if len(means) == len(sds) and len(means) >= 1:
            nds = [NormDist(0, 0) for _ in range(len(means))]
            for i in range(len(means)):
                nds[i].mean = means[i]
                nds[i].sd = sds[i]
            return nds
        else:
            raise ValueError('means and sds must be the same length and not empty.')

    @staticmethod
    def average_cells(nd_cells):
        if not isinstance(nd_cells, list) or not all(isinstance(cell, list) for cell in nd_cells):
            raise ValueError('Must be a list of lists')

        size1 = np.shape(nd_cells[0])
        ok = True
        for i in range(1, len(nd_cells)):
            if not all(isinstance(nd_cells[i], NormDist) for nd in nd_cells[i]):
                raise ValueError('each list must contain an array of NormDists.')

            sizen = np.shape(nd_cells[i])
            if len(sizen) != len(size1) or not all(sizen == size1):
                ok = False
                break

        if not ok:
            raise ValueError('All the arrays in the list must be the same size.')

        len_nd = np.prod(size1)
        cs = NormDist.init(np.zeros(len_nd), np.zeros(len_nd))
        as_nd = NormDist.init(np.zeros(len(nd_cells)), np.zeros(len(nd_cells)))
        for i in range(len(cs)):
            for j in range(len(nd_cells)):
                as_nd[j] = nd_cells[j][i]
            cs[i] = NormDist.average(as_nd)
        cs = np.reshape(cs, size1)
        return cs

    @staticmethod
    def sum_cells(nd_cells):
        if not isinstance(nd_cells, list) or not all(isinstance(cell, list) for cell in nd_cells):
            raise ValueError('Must be a list of lists')

        size1 = np.shape(nd_cells[0])
        ok = True
        for i in range(1, len(nd_cells)):
            if not all(isinstance(nd_cells[i], NormDist) for nd in nd_cells[i]):
                raise ValueError('each list must contain an array of NormDists.')

            sizen = np.shape(nd_cells[i])
            if len(sizen) != len(size1) or not all(sizen == size1):
                ok = False
                break

        if not ok:
            raise ValueError('All the arrays in the list must be the same size.')

        len_nd = np.prod(size1)
        cs = NormDist.init(np.zeros(len_nd), np.zeros(len_nd))
        as_nd = NormDist.init(np.zeros(len(nd_cells)), np.zeros(len(nd_cells)))
        for i in range(len(cs)):
            for j in range(len(nd_cells)):
                as_nd[j] = nd_cells[j][i]
            cs[i] = NormDist.sum(as_nd)
        cs = np.reshape(cs, size1)
        return cs

    @staticmethod
    def plus(a, b):
        if np.shape(a) != np.shape(b):
            raise ValueError('arrays must be the same size.')

        a_mean = [nd.mean for nd in a]
        b_mean = [nd.mean for nd in b]
        a_sd = [nd.sd for nd in a]
        b_sd = [nd.sd for nd in b]

        c_mean = np.add(a_mean, b_mean)
        c_sd = np.sqrt(np.add(np.square(a_sd), np.square(b_sd)))

        return NormDist.init(c_mean, c_sd)

    @staticmethod
    def minus(a, b):
        if np.shape(a) != np.shape(b):
            raise ValueError('arrays must be the same size.')

        a_mean = [nd.mean for nd in a]
        b_mean = [nd.mean for nd in b]
        a_sd = [nd.sd for nd in a]
        b_sd = [nd.sd for nd in b]

        c_mean = np.subtract(a_mean, b_mean)
        c_sd = np.sqrt(np.add(np.square(a_sd), np.square(b_sd)))

        return NormDist.init(c_mean, c_sd)

    @staticmethod
    def times(a, b):
        if isinstance(a, NormDist) and isinstance(b, NormDist):
            if (a.mean / a.sd) < 5 or (b.mean / b.sd) < 5:
                # Basically, if we need the distribution to be far
                # away from zero for this to work. Otherwise, we
                # can't really approximate the convolution as a
                # normal distribution.
                print('Convolving Normal distributions with significant density close to zero. '
                      'Approximation to normal distribution will be poor.')

            c_mean = a.mean * b.mean
            v = a.sd**2 * b.sd**2 + a.sd**2 * b.mean**2 + b.sd**2 * a.mean**2
            c_sd = np.sqrt(v)
        elif isinstance(a, NormDist):
            if isinstance(b, (int, float)):
                c_mean = a.mean * b
                c_sd = a.sd * np.sqrt(abs(b))
            else:
                c_means = [nd.mean * b for nd in a]
                c_sds = [nd.sd * np.sqrt(abs(b)) for nd in a]
                return NormDist.init(c_means, c_sds)
        else:
            if isinstance(a, (int, float)):
                c_mean = b.mean * a
                c_sd = b.sd * np.sqrt(abs(a))
            else:
                c_means = [nd.mean * a for nd in b]
                c_sds = [nd.sd * np.sqrt(abs(a)) for nd in b]
                return NormDist.init(c_means, c_sds)

        return NormDist(c_mean, c_sd)

    def __mul__(self, other):
        return self.times(self, other)

    @staticmethod
    def rdivide(a, b):
        if isinstance(a, NormDist) and isinstance(b, (int, float)):
            c_mean = a.mean / b
            c_sd = a.sd / np.sqrt(abs(b))
            return NormDist(c_mean, c_sd)
        else:
            raise ValueError('right side must be numeric')

    @staticmethod
    def average(as_nd):
        c = NormDist.sum(as_nd)
        c.mean = c.mean / len(as_nd)
        c.sd = c.sd / np.sqrt(len(as_nd))
        return c

    @staticmethod
    def sum(as_nd):
        c_mean = np.sum([nd.mean for nd in as_nd])
        c_sd = np.sqrt(np.sum([np.square(nd.sd) for nd in as_nd]))
        return NormDist(c_mean, c_sd)

File: Matlab Code/Classes/NormDist.m
classdef NormDist
    
   properties
      mean
      sd      
   end
   
   methods(Static = true)
       function nds = init(means, sds)
          if length(means) == length(sds) && length(means) >= 1
             a = NormDist(0, 0);
             nds(length(means)) = a;
             for i = 1:length(means)
                nds(i).mean = means(i);
                nds(i).sd = sds(i);
             end
          else
              error('means and sds must be the same length and not empty.');
          end
       end
       
       function cs = averageCells(ndCells)
          %Requires each cell in ndCells to be a NormDist array of the same
          %size. Returns the elementwise average across the cells.
          if ~iscell(ndCells)
              error('Must be a cell array')
          end
          
          size1 = size(ndCells{1});
          ok = true;
          for i = 2:length(ndCells)
              if ~isa(ndCells{i}, 'NormDist')
                  error('each cell must contain an array of NormDists.');
              end

              sizen = size(ndCells{i}); 
              if length(sizen) ~= length(size1)
                  ok = false;
                  break;
              end
              if ~all(sizen == size1)
                  ok = false;
                  break;
              end      
          end
          
          if (~ok)
              error('All the arrays in the cell array must be the same size.');
          end
          
          % We're good.       
          len = prod(size1);
          cs = NormDist.init(zeros(len, 1), zeros(len, 1));
          as = NormDist.init(zeros(length(ndCells), 1), zeros(length(ndCells), 1));
          for i = 1:length(cs)
             for j = 1:length(ndCells)
                 as(j) = ndCells{j}(i);
             end
             cs(i) = average(as);
          end
          cs = reshape(cs, size1);
       end
       
       function cs = sumCells(ndCells)
          %Requires each cell in ndCells to be a NormDist array of the same
          %size. Returns the elementwise average across the cells.
          if ~iscell(ndCells)
              error('Must be a cell array')
          end
          
          size1 = size(ndCells{1});
          ok = true;
          for i = 2:length(ndCells)
              if ~isa(ndCells{i}, 'NormDist')
                  error('each cell must contain an array of NormDists.');
              end

              sizen = size(ndCells{i}); 
              if length(sizen) ~= length(size1)
                  ok = false;
                  break;
              end
              if ~all(sizen == size1)
                  ok = false;
                  break;
              end      
          end
          
          if (~ok)
              error('All the arrays in the cell array must be the same size.');
          end
          
          % We're good.       
          len = prod(size1);
          cs = NormDist.init(zeros(len, 1), zeros(len, 1));
          as = NormDist.init(zeros(length(ndCells), 1), zeros(length(ndCells), 1));
          for i = 1:length(cs)
             for j = 1:length(ndCells)
                 as(j) = ndCells{j}(i);
             end
             cs(i) = sum(as);
          end
          cs = reshape(cs, size1);
       end
       
   end
   
   methods
       
       function obj = NormDist(mean, sd)           
           if (nargin < 2)
               if (nargin == 1)
                   if (isfield(mean, 'mean') && isfield(mean, 'sd'))
                      % Weird case that was cropping up. Thanks Matlab.
                      obj.mean = mean.mean;
                      obj.sd = mean.sd;
                      return;
                   else
                      error('Can''t supply just one argument to NormDist constructor unless it''s a NormDist to copy.');
                   end
               end
               mean = 0;
               sd = 0;
           end
          if isnumeric(mean) && isnumeric(sd)
              obj.mean = mean;
              obj.sd = sd;
          else
              error('Can only create NormDist with numeric input.')
          end
       end
       
       function c = plus(a, b)
           if ~all(size(a) == size(b))
               error('arrays must be the same size.');
           end
           aMean = [a.mean];
           bMean = [b.mean];
           aSD = [a.sd];
           bSD = [b.sd];
           cMean = aMean + bMean;
           cSD = sqrt(aSD.^2 + bSD.^2);
           c = NormDist.init(cMean, cSD);
       end
       
       function c = minus(a, b)           
           if ~all(size(a) == size(b))
               error('arrays must be the same size.');
           end
           aMean = [a.mean];
           bMean = [b.mean];
           aSD = [a.sd];
           bSD = [b.sd];
           cMean = aMean - bMean;
           cSD = sqrt(aSD.^2 + bSD.^2);
           c = NormDist.init(cMean, cSD);
       end
       
       function c = times(a, b)
           if isa(a, 'NormDist') && isa(b, 'NormDist')
               if ((a.mean / a.sd) < 5 || (b.mean / b.sd) < 5)
                   % Basically, if we need the distribution to be far
                   % away from zero for this to work. Otherwise we
                   % can't really approximate the convolution as a
                   % normal distribution.
                   warning('Convolving Normal distributions with significant density close to zero. Approximation to normal distribution will be poor.');
               end                   
               c.mean = a.mean * b.mean;
               v = a.sd^2 * b.sd^2 + a.sd^2 * b.mean^2 + b.sd^2 * a.mean^2;
               c.sd = sqrt(v);
           elseif isa(a, 'NormDist')
               cms = [a.mean] .* b;
               csds = [a.sd] .* sqrt(abs(b));
               c = NormDist.init(cms, csds);
%               c.mean = a.mean .* b;
%               c.sd = a.sd .* sqrt(b);
           else
               cms = [b.mean] .* a;
               csds = [b.sd] .* sqrt(abs(a));
               c = NormDist.init(cms, csds);
              
%               c.mean = b.mean .* a;
%               c.sd = b.sd .* sqrt(a);
           end
       end
      
       function c = mtimes(a, b)
           if isa(a, 'NormDist') && isa(b, 'NormDist')               
               if length(a) == 1 && length(b) == 1
                   if ((a.mean / a.sd) < 5 || (b.mean / b.sd) < 5)
                       % Basically, if we need the distribution to be far
                       % away from zero for this to work. Otherwise we
                       % can't really approximate the convolution as a
                       % normal distribution.
                       warning('Convolving Normal distributions with significant density close to zero. Approximation to normal distribution will be poor.');
                   end
                   
                   c.mean = a.mean * b.mean;
                   v = a.sd^2 * b.sd^2 + a.sd^2 * b.mean^2 + b.sd^2 * a.mean^2;
                   c.sd = sqrt(v);
               else
                   error('mtimes only works on scalar NormDists');
               end
           elseif isa(a, 'NormDist')
               if length(b) == 1
                   if length(a) == 1
                       c.mean = a.mean * b;
                       c.sd = a.sd * sqrt(abs(b));                       
                   else
                       cmeans = [a.mean] * b;
                       csds = [a.sd] * sqrt(abs(b));
                       c = NormDist.init(cmeans, csds);
%                       c.mean = a.mean * b;
%                       c.sd = a.sd * sqrt(b);                                              
                   end
               else
                   error('mtimes only works on scalar NormDists');
               end
           else
               if (length(a) == 1)
                   if length(b) == 1
                       c.mean = b.mean * a;
                       c.sd = b.sd * sqrt(abs(a));
                   else
                       cmeans = [b.mean] * a;
                       csds = [b.sd] * sqrt(abs(a));
                       c = NormDist.init(cmeans, csds);
                   end
               else
                   error('mtimes only works on scalar NormDists');                   
               end
           end
       end
       
       function c = rdivide(a, b)
           if isa(a, 'NormDist') && isnumeric(b)
               c.mean = a.mean / b;
               c.sd = a.sd / sqrt(abs(b));
           else
               error('right side must be numeric')
           end
       end
       
       function c = average(as)           
           c = sum(as);
           c.mean = c.mean / length(as);
           c.sd = c.sd / sqrt(length(as));
           % It's divided by the sqrt of the cells because if we had used
           % the full equation rather than the sum, we'd have divided by
           % the number of items inside the sqaure root. Since we didn't
           % we'll 'put it back in' here by sqrting it.
       end
       
       function c = sum(as)
           c.mean = sum([as.mean]);
           c.sd = sqrt(sum([as.sd].*[as.sd]));
       end
       
   end
   
end